### ISSUE TITLE
Add Tau and Pau to System.Math class

### ISSUE THREAD

**Issue by aaronfranke** [dotnet/corefx#26352]
The System.Math class currently has values for `PI`, and `[Theory]` for π, π/2, π/4, [etc](https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/tests/System/Math.cs), but there are a few other useful circle constants that can be useful to have that are missing. Namely:

* Tau, which is 2π, or about `6.28318530718`. In radians, all the way around the circle.

* [Pau](https://xkcd.com/1292/), which is 1.5π, or about `4.71238898038`. In radians, 3/4 of the way around the circle.

The names don't matter too much, though they're simple to list and remember. Also, values such as π/2 should be given a variable. I would like to be able to call `Math.TAU`, `Math.PAU`, `Math.PI`, `Math.HALFPI`, etc, or similar names, from within my program. 

Obviously I could just create my own helper class for this, but it would be better if it was implemented in .NET and C#, available for all users of .NET and C#.

**Comment by jnm2** [comment]
[Wait a minute...](https://xkcd.com/927/) tsk tsk 😜

**Comment by tannergooding** [comment]
Only `Math.Pi` and `Math.E` are exposed constants. The others are special test inputs/outputs (only defined in the unit test library) that are used to validate the math functions return the way expected value for some "well-know" values.

**Comment by aaronfranke** [comment]
Why not have more exposed constants? They are useful, after all.

**Comment by jnm2** [comment]
Are they? Let me turn that around and ask, why? Is it super common to maintain your own tau and half-pi constants? That's new to me.

**Comment by tannergooding** [comment]
I don't have any specific preference for or against exposing them. There are some libm implementations that expose similar constants, but they also haven't been exposed as part of the "public surface area" in the actual C/C++ standards.

They do represent core values that various math libraries may need to consume/produce; However, they are also just constants with well-defined values, so it isn't difficult for any particular library to define (and has no impact on the IL generated outside of having a member you can discover via reflection).

**Comment by aaronfranke** [comment]
I do know it's a joke, but it's still a simple name to use. Would you rather call the variable `Math.THREEHALVESPI` or `Math.THREEQUARTERSTAU`?

If they're simple to define in any library, why not define them in this library?

**Comment by tannergooding** [comment]
> If they're simple to define in any library, why not define them in this library?

1. I don't think anyone has proposed it yet.
2. There are many more (IMO) useful APIs to add to CoreFX, that provide significant benefit and that are not trivial to implement (this includes APIs in the System.Math library)

**Comment by tannergooding** [comment]
@aaronfranke, if you do want to "officially" propose CoreFX add the APIs, you should follow the steps [here](https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/api-review-process.md#steps) and submit a "formal" API proposal (probably by just updating the OP, rather than opening a new issue).

**Comment by karelz** [comment]
How common are these constants? How often are they used by System.Math users?
The names should be something everyone recognizes (discoverability), otherwise their value goes significantly down.

> why not define them in this library

We don't add every API we can, but only APIs, which are useful for a wider developer audience and are truly used / useful in real-world scenarios. Otherwise we would drown in lots of marginal APIs - making it harder for developers to choose which APIs to use. Everything comes at a cost - general API surface ease-of-use in this case.

**Comment by JonHanna** [comment]
> How common are these constants? 

Tau is not very commonly used, but it has its advocates (though the symbol has also been used for ½π and for the golden ratio instead of φ). Pau was invented by Randal Munroe to make fun of people who argue about whether or not we should use Tau.

**Comment by tannergooding** [comment]
@karelz, they are fairly core mathematical constants that are used in multiple places (we use many of them for validating correctness in our Math APIs, and I've seen a few scattered about in both managed and native land for CoreFX/CoreCLR). They are generally defined to account for minor inaccuracies in the floating-point format (e.g. `2 * Math.PI` may be slightly different than `Math.TwoPi`)

However, as indicated above, since these are well-defined constants, it is also trivial to add them to your own code base (but equally as trivial to add them to CoreFX).

LibM defines the following: https://www.gnu.org/software/libc/manual/html_node/Mathematical-Constants.html#Mathematical-Constants. If we were to add any constants, I would think they should be these

I would also vote we use names like `TwoPi`, rather than `Tau`.

**Comment by jnm2** [comment]
In theory, TwoPi and HalfPi have the exact same precision but the exponent is incremented or decremented, right?

**Comment by tannergooding** [comment]
> In theory, TwoPi and HalfPi have the exact same precision as PI but the exponent is incremented or decremented, right?

Basically, yes.
```C#
const double pi_d      = 3.1415926535897932;    // 0x400921FB54442D18
const double two_pi_d  = 6.2831853071795865;    // 0x401921FB54442D18
const double half_pi_d = 1.5707963267948966;    // 0x3FF921FB54442D18

const float pi_f       = 3.14159265f;           // 0x40490FDB
const float two_pi_f   = 6.28318531f;           // 0x40C90FDB
const float half_pi_f  = 1.57079633f;           // 0x3FC90FDB
```

**Comment by tannergooding** [comment]
And, for the default rounding mode, `Math.PI * 2` is exactly the same as `Math.TwoPi`. But it may not always be exactly the same in all rounding modes and other more complex constants may differ.

**Comment by jnm2** [comment]
In default rounding mode, how is it possible for the most precise representations *not* to have the exact same mantissa no matter how complex the constant is?

**Comment by aaronfranke** [comment]
There is also the matter of 1.5π, which is not a 2-power multiple of Pi.

**Comment by tannergooding** [comment]
> In default rounding mode, how is it possible for the most precise representations of ComplexConstant and 2 * ComplexConstant not to have the exact same mantissa no matter how complex the constant is?

Denormal numbers, where the exponent is zero 😄

**Comment by jnm2** [comment]
Oh, gotcha. But we aren't talking about adding constants that small, right?

**Comment by tannergooding** [comment]
> Oh, gotcha. But we aren't talking about adding constants that small, right?

No. But there are other multipliers (`1.5 * PI`) and more complex constants (`Log2(E)`) whose "exact" result may differ from the result returned by computing it in managed code (I haven't actually gone and checked which).

**Comment by jnm2** [comment]
Yes. I would expect actual multiplication to cause that. Though in the particular case of 0.75*pi, I'm getting 0x4002D97C7F3321D2 no matter how I calculate it as well as when letting the compiler pick the representation given a literal with more than enough digits. So the only advantage in adding any of these eighths increments of a circle would have to be as Karel said– they have to be of routine interest to people looking through the API.

**Comment by CodeBendor** [comment]
There are numerous constants available in math.h / cmath.h... https://msdn.microsoft.com/en-us/library/4hwaceh6.aspx when you define _USE_MATH_DEFINES for c/c++.  Seems like it'd be pretty simple to include these in the .Net System.Math and they are obviously useful.
